Source/WebCore/dom/ViewTransition.cpp

```diff
@@ -0,0 +1,268 @@
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ViewTransition.h"
+
+#include "Document.h"
+#include "JSDOMPromise.h"
+#include "JSDOMPromiseDeferred.h"
+
+namespace WebCore {
+
+static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(Document& document)
+{
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(document.globalObject());
+    JSC::JSLockHolder lock(globalObject.vm());
+    auto deferredPromise = DeferredPromise::create(globalObject);
+    auto domPromise = DOMPromise::create(globalObject, *JSC::jsCast<JSC::JSPromise*>(deferredPromise->promise()));
+    return { WTFMove(domPromise), deferredPromise.releaseNonNull() };
+}
+
+ViewTransition::ViewTransition(Document& document, RefPtr<ViewTransitionUpdateCallback>&& updateCallback)
+    : m_document(document)
+    , m_updateCallback(WTFMove(updateCallback))
+    , m_ready(createPromiseAndWrapper(document))
+    , m_updateCallbackDone(createPromiseAndWrapper(document))
+    , m_finished(createPromiseAndWrapper(document))
+{
+}
+
+ViewTransition::~ViewTransition() = default;
+
+Ref<ViewTransition> ViewTransition::create(Document& document, RefPtr<ViewTransitionUpdateCallback>&& updateCallback)
+{
+    return adoptRef(*new ViewTransition(document, WTFMove(updateCallback)));
+}
+
+DOMPromise& ViewTransition::ready()
+{
+    return m_ready.first.get();
+}
+
+DOMPromise& ViewTransition::updateCallbackDone()
+{
+    return m_updateCallbackDone.first.get();
+}
+
+DOMPromise& ViewTransition::finished()
+{
+    return m_finished.first.get();
+}
+
+// https://drafts.csswg.org/css-view-transitions/#skip-the-view-transition
+void ViewTransition::skipViewTransition(ExceptionOr<JSC::JSValue>&& reason)
+{
+    if (!m_document)
+        return;
+
+    Ref protectedThis { *this };
```

Comment:
No, we don't do protectors anymore. It is up to the caller of this function to protect `this`.